File: data-structures/lists/c-sharp/nested_lists_add.cs
/*
Raspberry Pi Foundation
Developed as part of Ada Computer Science 
Usage licensed under CC BY-NC-SA 4.0

Note: This file is designed to be copied out and compiled on your machine.
In order for it to compile properly you need to ensure that the project name is the same as the "namespace" in this file.
To run this file you need to:
1. Copy the contents
2. Paste them into the C# IDE of your choice (Visual Studio, for example)
3. Change the namespace to match your project (if necessary)
4. Compile the program
5. Run the program
*/

using System;
using System.Collections.Generic;

namespace AdaCodeSamples {
    class NestedLists {
        // The Main method is the entry point for all C# programs
        public static void Main() {
            // Declare and initialise the nested list of astronauts
            List<List<string>> allAstronauts = new List<List<string>>() {
                new List<string>() { "ISS", "Sergey Prokopyev", "176" },
                new List<string>() { "ISS", "Dmitry Petelin", "176" },
                new List<string>() { "ISS", "Frank Rubio", "176" },
                new List<string>() { "Tiangong", "Fei Junlong", "112" },
                new List<string>() { "Tiangong", "Deng Qingming", "107" },
                new List<string>() { "Tiangong", "Zhang Lu", "107" },
                new List<string>() { "ISS", "Stephen Bowen", "54" },
                new List<string>() { "ISS", "Warren Hoburg", "14" },
                new List<string>() { "ISS", "Sultan Al Neyadi", "14" },
                new List<string>() { "ISS", "Andrey Fedyaev", "14" }
            };

            // Ask the user to input the new astronaut's details
            Console.WriteLine("Enter the spacecraft's name:");
            string spacecraft = Console.ReadLine();

            Console.WriteLine("Enter the astronaut's full name:");
            string name = Console.ReadLine();

            Console.WriteLine("Enter the current number of days in space:");
            string days = Console.ReadLine();

            // Add the new astronaut to the list
            AddNewAstronaut(allAstronauts, spacecraft, name, days);

            // Output the details of the new astronaut
            int lastIndex = allAstronauts.Count - 1;
            List<string> astronaut = allAstronauts[lastIndex];
            DisplayAstronautDetails(astronaut);
        }

        // Add a new astronaut list to the end of the nested list
        public static void AddNewAstronaut(List<List<string>> allAstronauts,
            string givenSpacecraft, string givenName, string givenDays) {

            // Create a new list for the astronaut
            List<string> newAstronaut = new List<string>() {
                givenSpacecraft, givenName, givenDays
            };

            // Add the new astronaut list to the end of the nested list
            allAstronauts.Add(newAstronaut);
        }

        // Display the values from the given astronaut list
        public static void DisplayAstronautDetails(List<string> astronaut) {
            Console.WriteLine("\nAstronaut details:");

            // Repeat for each value in the list
            foreach (string value in astronaut) {
                Console.WriteLine(value);
            }
        }
    }
}

File: data-structures/lists/visual-basic/nested_lists_add.vb
' Raspberry Pi Foundation
' Developed as part of Ada Computer Science
' Usage licensed under CC BY-NC-SA 4.0
' Note: This file is designed to be copied out and compiled on your machine.
' To run this file you need to:
' 1. Copy the contents
' 2. Paste them into the Visual Basic IDE of your choice (Visual Studio, for example)
' 3. Compile the program
' 4. Run the program

Module Program

    ' The Main subroutine is the default entry point for a VB program
    Sub Main()
        ' Declare and initialise the astronaut list
        Dim allAstronauts As New List(Of List(Of String))({
            New List(Of String)({"ISS", "Sergey Prokopyev", "176"}),
            New List(Of String)({"ISS", "Dmitry Petelin", "176"}),
            New List(Of String)({"ISS", "Frank Rubio", "176"}),
            New List(Of String)({"Tiangong", "Fei Junlong", "112"}),
            New List(Of String)({"Tiangong", "Deng Qingming", "107"}),
            New List(Of String)({"Tiangong", "Zhang Lu", "107"}),
            New List(Of String)({"ISS", "Stephen Bowen", "54"}),
            New List(Of String)({"ISS", "Warren Hoburg", "14"}),
            New List(Of String)({"ISS", "Sultan Al Neyadi", "14"}),
            New List(Of String)({"ISS", "Andrey Fedyaev", "14"})
        })

        ' Ask the user to input the new astronaut's details
        Console.WriteLine("Enter the spacecraft's name:")
        Dim spacecraft As String = Console.ReadLine()

        Console.WriteLine("Enter the astronaut's full name:")
        Dim name As String = Console.ReadLine()

        Console.WriteLine("Enter the current number of days in space:")
        Dim days As String = Console.ReadLine()

        ' Add the new astronaut to the list
        AddNewAstronaut(allAstronauts, spacecraft, name, days)

        ' Output the details of the new astronaut
        Dim lastIndex As Integer = allAstronauts.Count - 1
        Dim astronaut As List(Of String) = allAstronauts(lastIndex)
        DisplayAstronautDetails(astronaut)
    End Sub

    ' Add a new astronaut list to the end of the nested list
    Sub AddNewAstronaut(ByVal allAstronauts As List(Of List(Of String)), ByVal givenSpacecraft As String, ByVal givenName As String, ByVal givenDays As String)

        ' Create a new list for the astronaut
        Dim newAstronaut As New List(Of String)({
            givenSpacecraft,
            givenName,
            givenDays
        })

        ' Add the new astronaut list to the end of the nested list
        allAstronauts.Add(newAstronaut)
    End Sub

    ' Display the values from the given astronaut list
    Sub DisplayAstronautDetails(ByVal astronaut As List(Of String))
        Console.WriteLine(vbLf & "Astronaut details:")

        ' Repeat for each value in the list
        For Each value In astronaut
            Console.WriteLine(value)
        Next
    End Sub
End Module